### PR TITLE
Restore downloads directory before downloading

### PR DESCRIPTION
**Pull request by michalpristas**
## What does this PR do?

During upgrade we cleanup `downloads` directory after ourselves. 
This makes troubles when upgrade is fine or rolled back, and then next upgrade needs to happen. 
Downloads directory does not exist and upgrade fails

Related PR here: https://github.com/elastic/elastic-agent/pull/752

## Why is it important?

Running multiple upgrades not possible

Logs from an example failure

```json
{"log.level":"info","@timestamp":"2023-01-13T10:03:54.035+0100","log.origin":{"file.name":"upgrade/upgrade.go","file.line":111},"message":"Upgrading agent","version":"8.6.0","source_uri":"","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2023-01-13T10:03:54.035+0100","log.origin":{"file.name":"upgrade/upgrade.go","file.line":132},"message":"Unable to clean downloads before update","error":{"message":"unable to read directory \"/opt/Elastic/Agent/data/elastic-agent-026915/downloads\": open /opt/Elastic/Agent/data/elastic-agent-026915/downloads: no such file or directory"},"downloads.path":"/opt/Elastic/Agent/data/elastic-agent-026915/downloads","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-13T10:03:54.035+0100","log.origin":{"file.name":"log/reporter.go","file.line":40},"message":"2023-01-13T10:03:54+01:00 - message: Application: [4535d1a9-c7b9-4913-86f8-808dabeb8fe1]: State changed to UPDATING: Update to version '8.6.0' started - type: 'STATE' - sub_type: 'UPDATING'","ecs.version":"1.6.0"}
{"log.level":"warn","@timestamp":"2023-01-13T10:03:54.035+0100","log.origin":{"file.name":"http/downloader.go","file.line":100},"message":"failed to cleanup : remove : no such file or directory","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2023-01-13T10:03:54.035+0100","log.origin":{"file.name":"upgrade/upgrade.go","file.line":149},"message":"Unable to remove file after verification failure","error":{"message":"unable to read directory \"/opt/Elastic/Agent/data/elastic-agent-026915/downloads\": open /opt/Elastic/Agent/data/elastic-agent-026915/downloads: no such file or directory"},"ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2023-01-13T10:03:55.196+0100","log.origin":{"file.name":"log/reporter.go","file.line":36},"message":"2023-01-13T10:03:55+01:00 - message: Application: [4535d1a9-c7b9-4913-86f8-808dabeb8fe1]: State changed to FAILED: failed upgrade of agent binary: 2 errors occurred:\n\t* package '/opt/Elastic/Agent/data/elastic-agent-026915/downloads/elastic-agent-8.6.0-linux-x86_64.tar.gz' not found: open /opt/Elastic/Agent/data/elastic-agent-026915/downloads/elastic-agent-8.6.0-linux-x86_64.tar.gz: no such file or directory\n\t* creating package file failed: open /opt/Elastic/Agent/data/elastic-agent-026915/downloads/elastic-agent-8.6.0-linux-x86_64.tar.gz: no such file or directory\n\n - type: 'ERROR' - sub_type: 'FAILED'","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2023-01-13T10:03:55.196+0100","log.origin":{"file.name":"handlers/handler_action_upgrade.go","file.line":45},"message":"Upgrade action failed","error":{"message":"failed upgrade of agent binary: 2 errors occurred:\n\t* package '/opt/Elastic/Agent/data/elastic-agent-026915/downloads/elastic-agent-8.6.0-linux-x86_64.tar.gz' not found: open /opt/Elastic/Agent/data/elastic-agent-026915/downloads/elastic-agent-8.6.0-linux-x86_64.tar.gz: no such file or directory\n\t* creating package file failed: open /opt/Elastic/Agent/data/elastic-agent-026915/downloads/elastic-agent-8.6.0-linux-x86_64.tar.gz: no such file or directory\n\n"},"action.version":"8.6.0","action.source_uri":"","action.id":"cb140906-3dee-4948-89d8-3757c3d66c34","action.start_timeError":"json: unsupported type: func() (time.Time, error)","action.expiration":"2023-01-13T10:33:32.339Z","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2023-01-13T10:03:55.197+0100","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":204},"message":"failed to dispatch actions, error: failed upgrade of agent binary: 2 errors occurred:\n\t* package '/opt/Elastic/Agent/data/elastic-agent-026915/downloads/elastic-agent-8.6.0-linux-x86_64.tar.gz' not found: open /opt/Elastic/Agent/data/elastic-agent-026915/downloads/elastic-agent-8.6.0-linux-x86_64.tar.gz: no such file or directory\n\t* creating package file failed: open /opt/Elastic/Agent/data/elastic-agent-026915/downloads/elastic-agent-8.6.0-linux-x86_64.tar.gz: no such file or directory\n\n","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2023-01-13T10:03:55.197+0100","log.origin":{"file.name":"status/reporter.go","file.line":326},"message":"Elastic Agent status changed to \"error\": \"component gateway-b0fe7321: failed to dispatch actions, error: failed upgrade of agent binary: 2 errors occurred:\\n\\t* package '/opt/Elastic/Agent/data/elastic-agent-026915/downloads/elastic-agent-8.6.0-linux-x86_64.tar.gz' not found: open /opt/Elastic/Agent/data/elastic-agent-026915/downloads/elastic-agent-8.6.0-linux-x86_64.tar.gz: no such file or directory\\n\\t* creating package file failed: open /opt/Elastic/Agent/data/elastic-agent-026915/downloads/elastic-agent-8.6.0-linux-x86_64.tar.gz: no such file or directory\\n\\n\"","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-13T10:04:13.688+0100","log.origin":{"file.name":"artifact/config.go","file.line":138},"message":"Source URI changed from \"https://artifacts.elastic.co/downloads/\" to \"https://artifacts.elastic.co/downloads/\"","ecs.version":"1.6.0"}
```

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)

Not fixing an actual issue, just one of the issues related to upgrade to 8.6

### CODE REVIEW

File: internal/pkg/agent/application/upgrade/step_download.go [hunk 1]
```diff
@@ -54,6 +57,10 @@ func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI stri
 		return "", errors.New(err, "initiating fetcher")
 	}
 
+	if err := os.MkdirAll(paths.Downloads(), 0755); err != nil {
```
Comment (aleksmaus) [inline internal/pkg/agent/application/upgrade/step_download.go:60]: nit: do we really need 0755? why not 0750 ?

**Review by blakerouse — APPROVED** [review]
Looks good.

If we can lower the permissions as @aleksmaus mentioned that would be a plus+

**Comment by cmacknz** [comment]
> This makes troubles when upgrade is fine or rolled back, and then next upgrade needs to happen.
Downloads directory does not exist and upgrade fails

Just to clarify this, particularly the "upgrade is fine or rolled back" part, do we only hit this bug when an upgrade is tried and then rolled back?

**Comment by cmacknz** [comment]
Additional question, is there anyway to automatically remediate agents in this situation without requiring users to run mkdir 100s of times on all affected agent machines?

**Comment by michalpristas** [comment]
@cmacknz yes, after upgrade we cleanup download. so once you start another upgrade you fail.

**Comment by michalpristas** [comment]
also about remedy, this version will be ok without creating it. previous not.  this version creates it and ignores path during cleanup

**Comment by cmacknz** [comment]
We need QA validation that the following upgrade path is broken:

Install 8.5.3 -> Upgrade to 8.6.0 -> Upgrade to 8.7.0-SNAPSHOT. The upgrade to 8.6.1 should fail.

We should then be able to confirm that the following upgrade path works since it includes the fix from this PR:

Install 8.5.3 -> Upgrade to 8.6.0-SNAPSHOT -> Upgrade to 8.7.0-SNAPSHOT

We should create a regression test case for this situation as a result of this testing as well covering an upgrade from the last minor, to the current minor, to the snapshot version of the next minor release.

**Comment by jlind23** [comment]
@dikshachauhan-qasource @amolnater-qasource would you be able to look at Craig's comment above and test the mentioned upgrade path?

File: internal/pkg/agent/application/upgrade/step_download.go [hunk 2]
```diff
@@ -54,6 +57,10 @@ func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI stri
 		return "", errors.New(err, "initiating fetcher")
 	}
 
+	if err := os.MkdirAll(paths.Downloads(), 0750); err != nil {
+		return "", errors.New(err, fmt.Sprintf("failed to create download directory at %s", paths.Downloads()))
+	}
```
Comment (cmacknz) [inline internal/pkg/agent/application/upgrade/step_download.go:62]: @michalpristas the downloader implementation looks like it does this internally: 

https://github.com/elastic/elastic-agent/blob/b8553c59a5a7c0b1944eee96562892fc748b0668/internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go#L179-L183

What am I missing here?

**Comment by amolnater-qasource** [comment]
Hi Team,

Thank you for the updates.
We will be revalidating this PR once latest 8.7.0 SNAPSHOT is available and will add the same scenarios for regression.

Thanks!

File: internal/pkg/agent/application/upgrade/step_download.go (hunk 2 above)
Comment (cmacknz) [reply internal/pkg/agent/application/upgrade/step_download.go:62]: Looks like the change here may not have actually been necessary speaking with Michal.

**Comment by amolnater-qasource** [comment]
Hi Team, 
We have revalidated this PR on latest 8.7.0 BC1 kibana cloud environment and had below observations:
- Installed agent v8.5.3
- Upgraded agent v8.5.3> v8.6.0> v8.6.1 successfully.
- Upgraded v8.6.1>v8.7.0 successfully by adding agent binary.


Further we are unable to test this PR on snapshot build as Snapshot versions are not available under agent upgrade flyout, also reported under https://github.com/elastic/kibana/issues/139174
Could you please share if there's any workaround for the same?

**Build details:**
BUILD: 60614
COMMIT: d3b239d76aa04f073836f6100782134ac86887e2

**Screenshots:**






Please let us know if anything else is required from our end.
Thanks!

**Comment by cmacknz** [comment]
> Further we are unable to test this PR on snapshot build as Snapshot versions are not available under agent upgrade flyout, also reported under https://github.com/elastic/kibana/issues/139174
> Could you please share if there's any workaround for the same?

The snapshot versions should be available in specific cloud regions, for example GCP us-west-2. If the official release versions work I am not that concerned about testing the snapshots if this does allow you to test it.

**Pull request merged**